### PR TITLE
Issue 4531 - add bunny fonts

### LINKED ISSUE #5430: Some google fonts missing in maxiblock editor - grotesk types
### Description

The following appear missing in the google fonts list:

Space Grotesk
Hanken Grotesk
Schibsted Grotesk
Familjen Grotesk
Be Vietnam Pro
Bricolage Grotesque

### Step-by-step reproduction instructions

when using any typography panel in the block editor

### Screenshots, screen recording, code snippet

_No response_

### URL and/or Code editor

_No response_

### How to fix it

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by myroslavve**
# Description
In dashboard, added option to use bunny fonts instead of Google fonts:
image

Fixes #4531 

# How Has This Been Tested?
Checked in devtools that fonts load from fonts.bunny.net when enabled.
image
Also checked that local fonts work with bunny fonts enabled.

# Test checklist

**_ Front/Back Testing _**

-   [ ] Test editor/frontend.
-   [ ] Test local fonts.

# Checklist

-   [ ] My code follows the style guidelines of this project
-   [ ] I have performed a self-review of my code
-   [ ] I have commented my code, particularly in hard-to-understand areas
-   [ ] My changes generate no new warnings/errors
-   [ ] I have added/updated tests that prove my fix is effective or that my feature works
-   [ ] New and existing unit tests pass locally with my changes

## Summary by CodeRabbit

- **New Features**
  - Introduced a new setting to switch between Google Fonts and Bunny Fonts APIs for font loading.
  - Enhanced font loading behavior with improved URL construction based on font weights and styles.

- **Refactor**
  - Improved readability and maintainability of font-related methods and functions.

- **Bug Fixes**
  - Ensured font settings are updated correctly during installation.

### CODE REVIEW

File: core/admin/class-maxi-dashboard.php [hunk 1]
```diff
@@ -485,9 +485,14 @@ public function maxi_blocks_settings()
             $content .= '</div>'; // maxi-dashboard_main-content_accordion-item
             $content .= $this->generate_item_header(__('Fonts and files', 'maxi-blocks'), false);
 
+
+            $description = '<h4>'.__('Use bunny.net fonts', 'maxi-blocks').'</h4>';
+            $description .= '<p>'.__('Use bunny.net CDN to serve fonts. GDPR compliant analogue to Google fonts', 'maxi-blocks').'</p>';
```
Comment (myroslavve) [inline core/admin/class-maxi-dashboard.php:490]: @ckp267 could you please check name/description for this in dashboard, maybe need to change or add anything.

**Comment by elzadj** [comment]
I think we also need to update the list of fonts we are using (/src/extensions/text/fonts/fonts.js) - it's old, and we are missing some fonts.

**Comment by kyrapieterse** [comment]
My chat with Svits - Let talk here

Why would you want local bunny fonts? Does it make it faster?

yes, local fonts are always faster
minus an external request
(ok, not always, if your server is shit, than external could be faster, but many people just don't want to depend on external services)

if google fonts goes down, half of internet shows Times New Roman

I agree if its working dont change it

**Comment by elzadj** [comment]
Or we can just always download local fonts from bunny, then 3 options is ok.
Maybe we don't need to let people choose where to download them from. Again, need to check the legal staff.

**Comment by ckp267** [comment]
### Updated Font Settings Proposal for MaxiBlocks Dashboard 
( agreed in DM's with @myroslavve @elzadj @kyrapieterse )

_BEGIN - Settings accordion text and options_

### Fonts and files

**Use Bunny Fonts**
Privacy-friendly, GDPR compliant. Global CDN for fast loading. Wide selection of fonts available.
[x] Use Bunny Fonts (Default: On)

[x] Serve Bunny Fonts from CDN (Default: On)
Fastest option. Uses external CDN. No local storage required.

[ ] Serve Bunny Fonts locally (Default: Off)
Privacy-focused. May impact server performance. Requires local storage.

**Google Fonts**
[ ] Use Google Fonts (Default: Off)
Extensive font selection. Potential privacy concerns when using Google's CDN.

[x] Serve Google Fonts locally (Default: On if Google Fonts is used)
Blocks Google tracking. May impact server performance. Requires local storage.

[ ] Serve from Google CDN (Default: Off)
"Fastest option. Uses Google's CDN. Potential privacy (GDPR) implications."

**General note**
Local storage: Download and serve font files from your WordPress uploads directory. Improves privacy but may affect performance. Unused fonts are periodically removed to save space.

_END - Settings accordion text and options_

### Implementation Notes:
* Respect user's previous settings to avoid overwriting customizations
 - New users: bunny external by default (as per defaults above)
 - **IMPORTANT!!** Old users: whatever they used before. We need to test this and confirm!

* **IMPORTANT!!** Only ONE top level option. Either bunny OR google. Cannot have both at same time.

* Update the font list in /src/extensions/text/fonts/fonts.js.

* Refactor class-maxi-local-fonts.php for improved efficiency.

* Test current implementation periodic cleanup of unused local font files - make sure it works

**Comment by myroslavve** [comment]
@Olekrut 
Need to check 
>  - New users: bunny external by default (as per defaults above)
> - IMPORTANT!! Old users: whatever they used before. We need to test this and confirm!

So when you install plugin, even without opening maxi dashboard, frontend should have bunny fonts, and if you switch to this branch from master should still be google fonts.
Also, check fonts on frontend in general please, since the code was refactored.

**Comment by myroslavve** [comment]
Also fixes #5430

**Review by Olekrut — CHANGES_REQUESTED** [review]
https://jam.dev/c/b7ac4e6f-dca5-4062-be38-ee945f4daa97
1) created page on master > google fonts
2) switch brunch to this > fonts changed to bunny

**Comment by elzadj** [comment]
> https://jam.dev/c/b7ac4e6f-dca5-4062-be38-ee945f4daa97
> 
> 1. created page on master > google fonts
> 2. switch brunch to this > fonts changed to bunny

@myroslavve could it be because the local env. already has bunny_fonts set to true? Need to figure out something here, at least for tests.

**Comment by myroslavve** [comment]
> https://jam.dev/c/b7ac4e6f-dca5-4062-be38-ee945f4daa97
> 
> 1. created page on master > google fonts
> 2. switch brunch to this > fonts changed to bunny

If you have bunny fonts enabled before switching to master, it will remain enabled, and same for disabled state. So seems to be working right.

https://github.com/maxi-blocks/maxi-blocks/assets/53875005/edd19761-0b23-48c0-b40d-cb0d0653e0a3

**Review by Olekrut — APPROVED** [review]

**Review by elzadj — CHANGES_REQUESTED** [review]
Great work! A couple of notes:
1. The font id on backed has spaces, if the font name has them. Ideally need to make it the same as on the frontend, like ```id='maxi-blocks-styles-font-roboto-css'``` 
Screenshot from 2024-07-12 15-34-51
2. It's absolutely not clear that you can switch to Google fonts (or to Bunny fonts) here, or what you are using right now. Maybe just write it near the option? "You are using: Google fonts" or "You are using: Bunny fonts". @ckp267 
screenshot-wordpress local-2024 07 12-15_49_15
screenshot-wordpress local-2024 07 12-16_01_07
3. **Local bunny and google fonts don't work on frontend** - the style.css and font files are missing from uploads/maxi folder.
screenshot-wordpress local-2024 07 12-16_07_52
@Olekrut please confirm, remove your wp-conent/uploads/maxi/fonts folder before activating the local option.
5. **Backend still loads the fonts from the external bunny site, if local option is enabled, is must be served from the local folder on backend too.**
6. Remove all commented out code, please.

**Review by Olekrut — CHANGES_REQUESTED** [review]
hmm, yes local fonts are not working correct, only after changing font for some text and updating, then everything is appearing

**Comment by myroslavve** [comment]
> 1. The font id on backed has spaces, if the font name has them. Ideally need to make it the same as on the frontend, like `id='maxi-blocks-styles-font-roboto-css'`
>    Screenshot from 2024-07-12 15-34-51

Fixed

> 2. It's absolutely not clear that you can switch to Google fonts (or to Bunny fonts) here, or what you are using right now. Maybe just write it near the option? "You are using: Google fonts" or "You are using: Bunny fonts". @ckp267
>    screenshot-wordpress local-2024 07 12-15_49_15
>    screenshot-wordpress local-2024 07 12-16_01_07

Added an indicator that shows which font provider is currently used, let me know if you have any other ideas to improve the labels
image


> 3. **Local bunny and google fonts don't work on frontend** - the style.css and font files are missing from uploads/maxi folder.
>    screenshot-wordpress local-2024 07 12-16_07_52

Fixed

> 4. **Backend still loads the fonts from the external bunny site, if local option is enabled, is must be served from the local folder on backend too.**

Doesn't work on master as well, I think it would be better to create a separate issue, because need to connect js and php code to save font files while in editor, could be complicated.

> 5. Remove all commented out code, please.

Done

**Review by Olekrut — APPROVED** [review]

**Review by elzadj — CHANGES_REQUESTED** [review]
> > 1. The font id on backed has spaces, if the font name has them. Ideally need to make it the same as on the frontend, like `id='maxi-blocks-styles-font-roboto-css'`
> >    Screenshot from 2024-07-12 15-34-51
> 
> Fixed
> 

 1. I still see 'font-Roboto+500+normal' - please make it with 'maxi-blocks' to identify with our plugin, that's going to help with future debugging. Also please replace special symbols with '-' and make it all lower case. You can add 'backend' or 'editor' too, if you want.

2. I think we need to move it on top, looks weird now. Something like 
Fonts and files
You are using .... right now
Options
But it's a question for @ckp267 and @kyrapieterse 

4. Yeah, but on master we use ```new FontFace```, so the font file itself is not directly visible in the page source code. Now we see the external link right away if we check the source code. Need to do it asap, after this one, or go back to the FontFace API.

**Comment by myroslavve** [comment]
> 1. I still see 'font-Roboto+500+normal' - please make it with 'maxi-blocks' to identify with our plugin, that's going to help with future debugging. Also please replace special symbols with '-' and make it all lower case. You can add 'backend' or 'editor' too, if you want.

Added 'maxi-blocks' to the id, not sure which special symbols to replace, since font names don't have any in them.

> 2. I think we need to move it on top, looks weird now. Something like
>    Fonts and files
>    You are using .... right now
>    Options
>    But it's a question for @ckp267 and @kyrapieterse

When it is on top it looks kind of weird, unless we add some styles
image
or
image
Looks a little better underneath the setting title:
image
so left it like this. Not sure if it is good enough though :shrug:

> 3. Yeah, but on master we use `new FontFace`, so the font file itself is not directly visible in the page source code. Now we see the external link right away if we check the source code. Need to do it asap, after this one, or go back to the FontFace API.

Okay, I will be working on the issue now, but in separate PR

**Comment by kyrapieterse** [comment]
Looks a little better underneath the setting title: 

looks ok. MVP - when Its merged I can take another look

**Comment by elzadj** [comment]
> Added 'maxi-blocks' to the id, not sure which special symbols to replace, since font names don't have any in them.

I meant "+" in the id. 

>ID and NAME tokens must begin with a letter ([A-Za-z]) and may be followed by any number of letters, digits ([0-9]), hyphens ("-"), underscores ("_"), colons (":"), and periods (".").

We can ignore it for now, and fix later, if we have any problems with it.

**Review by elzadj — APPROVED** [review]

**Pull request merged**